Supports super().__torch_dispatch__ with arguments list

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74509

Summary:
For THPModule_disable_torch_(dispatch|function),  converts list arguments to tuple before invoking PyObject_Call.

Fixes  #73933

Differential Revision: [D35034204](https://our.internmc.facebook.com/intern/diff/D35034204)

SherlockNoMad committed:
Supports super().__torch_dispatch__ with arguments list

Summary:
For THPModule_disable_torch_(dispatch|function),  converts list arguments to tuple before invoking PyObject_Call.

Fixes  #73933

Test Plan:

Reviewers:

Subscribers:

Tasks: 114830027

Tags:

[ghstack-poisoned]

SherlockNoMad commented:
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034204).

ezyang commented on torch/csrc/utils/disable_torch_function.cpp:
```diff
@@ -119,8 +119,9 @@ PyObject* THPModule_disable_torch_function(PyObject *self, PyObject *a) {
-  }
-  else {
+  } else if (PyList_CheckExact(args)) {
+    py_args = py::reinterpret_borrow<py::tuple>(PyList_AsTuple(args));
```
reinterpret_borrow here is wrong as PyList_AsTuple returns a new reference; if you only borrow it you will leak the allocated tuple here. https://github.com/pybind/pybind11/issues/1201 may be of interest.

SherlockNoMad replied:
ah... I need to use reinterpret_steal here, as a move semantics. Thanks a lot for the cheatsheet, it's way clearer than pybind's doc. :)

ezyang reviewed (changes requested):
Current implementation leaks PyObjects.

SherlockNoMad committed:
Update on "Supports super().__torch_dispatch__ with arguments list"


Summary:
For THPModule_disable_torch_(dispatch|function),  converts list arguments to tuple before invoking PyObject_Call.

Fixes  #73933

Differential Revision: [D35034204](https://our.internmc.facebook.com/intern/diff/D35034204)

[ghstack-poisoned]

SherlockNoMad commented:
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034204).

SherlockNoMad committed:
Update on "Supports super().__torch_dispatch__ with arguments list"


Summary:
For THPModule_disable_torch_(dispatch|function),  converts list arguments to tuple before invoking PyObject_Call.

Fixes  #73933

Differential Revision: [D35034204](https://our.internmc.facebook.com/intern/diff/D35034204)

[ghstack-poisoned]

SherlockNoMad commented:
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35034204).

SherlockNoMad commented:
@pytorchbot merge this please
